### PR TITLE
Feature: Serve static files Asynchronously

### LINKED ISSUE #674: Add file support in `HttpData`
**Is your feature request related to a problem? Please describe.**
Currently the only way to serve files is via `ZStream.fromFile`. This works, however is slow and uses a blocking thread pool.
```scala
Http.collect[Request] { 
 case Get -> !! / "file" => Response(data = HttpData.fromStream(ZStream.fromFile(Paths.get("README.md")))
}
```

**Describe the solution you'd like**
It would better to use a specialized constructor in HttpData that reads files efficiently.
```scala
Http.collect[Request] { 
 case Get -> !! / "file" => Response(data = HttpData.fromPath(Paths.get("README.md"))
}
```
Where `HttpData.fromPath` internally uses netty's `ChunkedFile` API.

- This solution will be significantly more performant since it's going to run on the same thread as the incoming request.
- Later the implementation can be used to build a static server eg: `Http.static("~/public")`
- Ideally the `content-size` should be set in the response header because files on disk typically have a finite size.

**Describe alternatives you've considered**
We create a new file reader that reads file in chunks using netty's `ChunkedFile` API and then use `HttpData.fromStream` to create an HttpData.
We need to still test how well it performs under load.

```scala
  final class ChunkedReader(file: ChunkedFile, alloc: ByteBufAllocator) {
    def read: ZIO[Any, Option[Throwable], ByteBuf] =
      ZIO.effectSuspendTotal {
        try {
          val data = file.readChunk(alloc)
          if (data == null) ZIO.fail(None)
          else UIO(data)
        } catch {
          case error: Throwable => ZIO.fail(Option(error))
        }
      }

    def toStream: ZStream[Clock, Throwable, ByteBuf] = ZStream.repeatEffectOption(read)
  }

  object ChunkedReader {
    def make(path: java.nio.file.Path, alloc: ByteBufAllocator = ByteBufAllocator.DEFAULT): ZIO[Any, Throwable, ZIOChunkedFile] = for {
      file  <- Task(path.toFile)
      cFile <- Task(new ChunkedFile(file, 128))
    } yield new ZIOChunkedFile(cFile, alloc)
  }

// Inside HttpData

  def fromFile(path: java.nio.file.Path): HttpData[Clock, Throwable] =
    HttpData.BinaryStream(ZStream.unwrap(ZIOChunkedFile.make(path).map(_.toStream)))
```

**Additional context**
- Will need to add a `ChunkedWriteHandler` in the pipeline if the response is a file
- Sample implementation: https://github.com/netty/netty/blob/4.1/example/src/main/java/io/netty/example/file/FileServerHandler.java

### PR DESCRIPTION
**Pull request by ashprakasan**
Fixes #674, #238
Some to-do items 
- [x] Implement `Http.fromFile` to create an Http app from the contents of a file.
- [x] Test a failing scenario with `toByteBuf` where file size is too big or handle this scenario elegantly

### CODE REVIEW

File: example/src/main/scala/example/FileStreaming.scala [hunk 1]
```diff
@@ -5,18 +5,31 @@ import zhttp.service.Server
...
   }
+  val filePathAsString = File.createTempFile("example/blahFile", ".avi").getAbsolutePath
+
+  def testFile = {
+    val tempFileName = File.createTempFile("example/blahTextFile", ".txt").getAbsolutePath
+    Files.write(
+      Paths.get(tempFileName),
+      "This is a test file created only for testing.\n".getBytes(StandardCharsets.UTF_8),
+    )
+  }
```
Comment (tusharmath) [inline example/src/main/scala/example/FileStreaming.scala]: Can add a file to resources and serve it directly instead of creating one in the example.

File: example/src/main/scala/example/FileStreaming.scala [hunk 2]
```diff
@@ -5,18 +5,31 @@ import zhttp.service.Server
...
 
   // Create HTTP route
   val app = Http.collect[Request] {
-    case Method.GET -> !! / "health" => Response.ok
-    case Method.GET -> !! / "file"   => Response(data = content)
+    case Method.GET -> !! / "health"   => Response.ok
+    case Method.GET -> !! / "file"     => Response(data = content)
+    case Method.GET -> !! / "withMime" => Response(data = HttpData.fromFile(Paths.get(filePathAsString)))
+    case Method.GET -> !! / "test"     => Response(data = HttpData.fromFile(testFile))
```
Comment (tusharmath) [inline example/src/main/scala/example/FileStreaming.scala]: Little bit of documentation on the new `fromFile` operator.

File: zio-http/src/main/scala/zhttp/http/HttpData.scala [hunk 3]
```diff
@@ -47,6 +50,17 @@ sealed trait HttpData[-R, +E] { self =>
     case HttpData.BinaryByteBuf(data)  => UIO(data)
     case HttpData.Empty                => UIO(Unpooled.EMPTY_BUFFER)
     case HttpData.BinaryStream(stream) => stream.fold(Unpooled.compositeBuffer())((c, b) => c.addComponent(b))
+    case HttpData.File(path)           =>
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpData.scala]: should read the whole file. Needs unit tests.

File: zio-http/src/main/scala/zhttp/http/Response.scala [hunk 4]
```diff
@@ -24,6 +25,16 @@ final case class Response[-R, +E] private (
   def addCookie(cookie: Cookie): Response[R, E] =
     self.copy(headers = self.getHeaders ++ Headers(HttpHeaderNames.SET_COOKIE.toString, cookie.encode))
 
+  def setMimeType: Response[R, E] = {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Response.scala]: can drop this method.

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 5]
```diff
@@ -60,8 +67,18 @@ private[zhttp] final case class Handler[R](
...
+    val jHeaders: HttpHeaders = res.getHeaders.encode
     if (res.attribute.serverTime) jHeaders.set(HttpHeaderNames.DATE, serverTime.refreshAndGet())
+
+    /**
+     * Sets MIME type in the response headers. This is only relevant in case of File transfers as browsers use the MIME
+     * type, not the file extension, to determine how to process a URL. <a href="MSDN
+     * Doc">https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type</a>
+     */
+    res.data match {
+      case HttpData.File(path) => jHeaders.set(HttpHeaderNames.CONTENT_TYPE.toString, Files.probeContentType(path))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala]:
```suggestion
      case HttpData.File(path) => jHeaders.set(HttpHeaderNames.CONTENT_TYPE.toString, Files.probeContentType(path))
```
```suggestion
      case HttpData.File(path) => jHeaders.set(HttpHeaderNames.CONTENT_TYPE, Files.probeContentType(path))
```

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 6]
```diff
@@ -244,4 +271,13 @@ private[zhttp] final case class Handler[R](
       _ <- ChannelFuture.unit(ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT))
     } yield ()
   }
+
+  private def unsafeWriteFileContent(path: java.nio.file.Path)(implicit ctx: ChannelHandlerContext) = {
+    import java.io.RandomAccessFile
+    val file       = new File(path.toString)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala]: should throw an 500 error if there is a failure. 
Add unit tests to handle error.

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 7]
```diff
@@ -244,4 +271,13 @@ private[zhttp] final case class Handler[R](
       _ <- ChannelFuture.unit(ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT))
     } yield ()
   }
+
+  private def unsafeWriteFileContent(path: java.nio.file.Path)(implicit ctx: ChannelHandlerContext) = {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala]: Bit of scala doc to explain other contributors what this method is doing.

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 8]
```diff
@@ -244,4 +271,13 @@ private[zhttp] final case class Handler[R](
...
+    import java.io.RandomAccessFile
+    val file       = new File(path.toString)
+    val raf        = new RandomAccessFile(file, "r")
+    val fileLength = raf.length()
+    ctx.write(new DefaultFileRegion(raf.getChannel, 0, fileLength))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala:227]: Can do it here or create a separate issue for handling range requests. Such that the range is not hardcoded to `0-fileLength`.

Comment (ashprakasan) [reply zio-http/src/main/scala/zhttp/service/Handler.scala:227]: I'll do this separately -  [#709](https://github.com/dream11/zio-http/issues/709)

File: zio-http/src/main/scala/zhttp/service/Handler.scala (hunk 6 above)
Comment (ashprakasan) [reply zio-http/src/main/scala/zhttp/service/Handler.scala]: Changed the method signature to address this. 
`HttpData.fromFile(file : java.io.File) = HttpData.File(file)`

File: example/src/main/scala/example/FileStreaming.scala [hunk 9]
```diff
@@ -13,10 +14,16 @@ object FileStreaming extends App {
     ZStream.fromFile(Paths.get("README.md"))
   }
 
+  // Read from file path directly
+  val videoFileContent = HttpData.fromFile(new File("src/main/resources/TestVideoFile.mp4"))
+  val textFileContent  = HttpData.fromFile(new File("src/main/resources/TestFile.txt"))
```
Comment (tusharmath) [inline example/src/main/scala/example/FileStreaming.scala:21]: can talk a bit about how the mime-types are automatically added.

Comment (tusharmath) [reply example/src/main/scala/example/FileStreaming.scala:21]: Talk a little about the differences between the two approaches.

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 10]
```diff
@@ -62,6 +64,17 @@ private[zhttp] final case class Handler[R](
...
     if (res.attribute.serverTime) jHeaders.set(HttpHeaderNames.DATE, serverTime.refreshAndGet())
+
+    /**
+     * Sets MIME type in the response headers. This is only relevant in case of File transfers as browsers use the MIME
+     * type, not the file extension, to determine how to process a URL. <a href="MSDN
+     * Doc">https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type</a>
+     */
+    res.data match {
+      case HttpData.File(file) =>
+        jHeaders.set(HttpHeaderNames.CONTENT_TYPE.toString, Files.probeContentType(Paths.get(file.toString)))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala]:
```suggestion
        jHeaders.set(HttpHeaderNames.CONTENT_TYPE, Files.probeContentType(Paths.get(file.toString)))
```

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 11]
```diff
@@ -62,6 +64,17 @@ private[zhttp] final case class Handler[R](
...
+      case _                   =>
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala]:
```suggestion
      case _                   => ()
```

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 12]
```diff
@@ -130,6 +131,11 @@ object ServerSpec extends HttpRunnableSpec(8088) {
         assertM(res)(equalTo(string))
       }
     } +
+      testM("data from file") {
+        val f: File = new File(getClass.getResource("/TestFile.txt").getPath)
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]:
```suggestion
        val file = new File(getClass.getResource("/TestFile.txt").getPath)
```

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 13]
```diff
@@ -130,6 +131,11 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
+        val res     = Http.data(HttpData.fromFile(f)).requestBodyAsString()
+        assertM(res)(equalTo("abc\nfoo"))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala:137]: Add test for `content-type`

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 14]
```diff
@@ -130,6 +131,11 @@ object ServerSpec extends HttpRunnableSpec(8088) {
         assertM(res)(equalTo(string))
       }
     } +
+      testM("data from file") {
+        val f: File = new File(getClass.getResource("/TestFile.txt").getPath)
+        val res     = Http.data(HttpData.fromFile(f)).requestBodyAsString()
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: Can do here or in a different PR — 

```suggestion
        val res     = Http.fromFile(file).requestBodyAsString()
```

File: build.sbt [hunk 15]
```diff
@@ -137,7 +137,7 @@ lazy val example = (project in file("./example"))
   .settings(publishSetting(false))
   .settings(
     fork                      := true,
-    Compile / run / mainClass := Option("example.HelloWorld"),
+    Compile / run / mainClass := Option("example.FileStreaming"),
```
Comment (tusharmath) [inline build.sbt]:
```suggestion
    Compile / run / mainClass := Option("example.HelloWorld"),
```

File: zio-http/src/main/scala/zhttp/http/HttpData.scala [hunk 16]
```diff
@@ -47,6 +49,12 @@ sealed trait HttpData[-R, +E] { self =>
     case HttpData.BinaryByteBuf(data)  => UIO(data)
     case HttpData.Empty                => UIO(Unpooled.EMPTY_BUFFER)
     case HttpData.BinaryStream(stream) => stream.fold(Unpooled.compositeBuffer())((c, b) => c.addComponent(b))
+    case HttpData.File(file)           =>
+      UIO {
+        val fileChannel: FileChannel = new FileInputStream(file).getChannel
+        val mappedByteBuffer         = fileChannel.map(FileChannel.MapMode.READ_ONLY, 0, file.length)
+        Unpooled.wrappedBuffer(mappedByteBuffer)
+      }
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpData.scala]: Need tests for this.

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 17]
```diff
@@ -55,17 +62,23 @@ private[zhttp] final case class Handler[R](
...
-
     // Identify if the server time should be set and update if required.
     if (res.attribute.serverTime) jResponse.headers().set(HttpHeaderNames.DATE, serverTime.refreshAndGet())
+
+    // Set MIME type in the response headers. This is only relevant in case of File transfers as browsers use the MIME
+    // type, not the file extension, to determine how to process a URL.<a href="MSDN
+    // Doc">https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type</a>
+    res.data match {
+      case HttpData.File(file) =>
+        jResponse.headers().set(HttpHeaderNames.CONTENT_TYPE, Files.probeContentType(file.toPath))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala]: Can move this to `Response.unsafeEncode`

File: zio-http/src/main/scala/zhttp/http/Response.scala [hunk 18]
```diff
@@ -76,13 +79,23 @@ final case class Response[-R, +E] private (
...
+      case HttpData.File(file)          =>
+        // Transfers the content of file channel to ByteBuf
+        val aFile: RandomAccessFile = new RandomAccessFile(file.toString, "r")
+        val inChannel: FileChannel  = aFile.getChannel
+        val fileSize: Long          = inChannel.size
+        val buffer: ByteBuffer      = ByteBuffer.allocate(fileSize.toInt)
+        inChannel.read(buffer)
+        inChannel.close()
+        aFile.close()
+        Unpooled.wrappedBuffer(buffer.flip)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Response.scala]: We should return `null` for file here.

File: zio-http/src/test/scala/zhttp/http/EncodeClientParamsSpec.scala [hunk 19]
```diff
@@ -9,11 +9,12 @@ import zio.test._
 
 object EncodeClientParamsSpec extends DefaultRunnableSpec with EncodeClientParams {
 
-  val anyClientParam: Gen[Random with Sized, Client.ClientParams] = HttpGen.clientParams(
+  val anyClientParam: Gen[Random with Sized, Client.ClientParams]         = HttpGen.clientParams(
     HttpGen.httpData(
       Gen.listOf(Gen.alphaNumericString),
     ),
   )
+  val clientParamForFileData: Gen[Random with Sized, Client.ClientParams] = HttpGen.clientParamsForFileHttpData()
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/EncodeClientParamsSpec.scala]: Can inline this value.

Comment (tusharmath) [reply zio-http/src/test/scala/zhttp/http/EncodeClientParamsSpec.scala:17]: On second thoughts, may be Gen is not required. We are always reading a file from `resources` dir.

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 20]
```diff
@@ -0,0 +1,15 @@
+package zhttp.http
+
+import zio.test.Assertion.equalTo
+import zio.test.{DefaultRunnableSpec, ZSpec, assertM}
+
+import java.io.File
+
+object HttpDataSpec extends DefaultRunnableSpec {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala:8]: May be add a `TODO` for to add tests for the rest of the HttpData types.

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+  val suite1 = suite("Test toByteBuf")(testM("HttpData.File") {
+    val file = new File(getClass.getResource("/TestFile.txt").getPath)
+    val res  = HttpData.fromFile(file).toByteBuf.map(_.toString(HTTP_CHARSET))
+    assertM(res)(equalTo("abc\nfoo"))
+  })
+  override def spec: ZSpec[_root_.zio.test.environment.TestEnvironment, Any] = suite("HttpDataSpec")(suite1)
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]:
```suggestion
  override def spec = suite("HttpDataSpec")(suite1)
```

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 22]
```diff
@@ -0,0 +1,15 @@
+package zhttp.http
+
+import zio.test.Assertion.equalTo
+import zio.test.{DefaultRunnableSpec, ZSpec, assertM}
+
+import java.io.File
+
+object HttpDataSpec extends DefaultRunnableSpec {
+  val suite1 = suite("Test toByteBuf")(testM("HttpData.File") {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]: can inline `suite1`

File: zio-http/src/test/scala/zhttp/internal/HttpGen.scala [hunk 23]
```diff
@@ -17,6 +19,15 @@ object HttpGen {
       data    <- dataGen
     } yield ClientParams(method, url, headers, data)
 
+  def clientParamsForFileHttpData() = {
+    val file = new File(getClass.getResource("/TestFile.txt").getPath)
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/internal/HttpGen.scala]: Should wrap `file` in a Task

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 24]
```diff
@@ -130,6 +132,21 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
+        assertM(res)(equalTo("abc\nfoo"))
+      } +
+      testM("content-type header on file response") {
+        val file = new File(getClass.getResource("/TestFile.txt").getPath)
+        val res  =
+          Http
+            .fromData(HttpData.fromFile(file))
+            .request()
+            .map(_.getHeaderValue(HttpHeaderNames.CONTENT_TYPE))
+            .map(_.get)
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]:
```suggestion
            .requestHeaderValueByName(Headers.Literals.Name.CONTENT_TYPE)
```

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 25]
```diff
@@ -130,6 +132,21 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
+        assertM(res)(equalTo("text/plain"))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala:146]:
```suggestion
        assertM(res)(equalTo(Headers.Literals.Value.TextPlain))
```

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 26]
```diff
@@ -194,7 +209,7 @@ object ServerSpec extends HttpRunnableSpec(8088) {
   override def spec = {
     suiteM("Server") {
       app.as(List(staticAppSpec, dynamicAppSpec, responseSpec, requestSpec)).useNow
-    }.provideCustomLayerShared(env) @@ timeout(30 seconds)
+    }.provideCustomLayerShared(env) @@ timeout(30 seconds) @@ sequential
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]:
```suggestion
    }.provideCustomLayerShared(env) @@ timeout(30 seconds)
```

File: zio-http/src/main/scala/zhttp/http/HttpData.scala [hunk 27]
```diff
@@ -33,12 +38,31 @@ sealed trait HttpData[-R, +E] { self =>
...
 
-  def toByteBuf: ZIO[R, E, ByteBuf] = self match {
-    case HttpData.Text(text, charset)  => UIO(Unpooled.copiedBuffer(text, charset))
-    case HttpData.BinaryChunk(data)    => UIO(Unpooled.copiedBuffer(data.toArray))
-    case HttpData.BinaryByteBuf(data)  => UIO(data)
-    case HttpData.Empty                => UIO(Unpooled.EMPTY_BUFFER)
-    case HttpData.BinaryStream(stream) => stream.fold(Unpooled.compositeBuffer())((c, b) => c.addComponent(b))
+  // Added for implicit conversion of `ZIO[R,E,ByteBuf]` to `ZIO[R, Throwable,Bytebuf]` in `toByteBuf`, may not be needed in scala3
+  implicit def convertZKT[F[-_, +_, +_], W, EA, EB, C](fa: F[W, EA, C])(implicit ev: EA <:< EB): F[W, EB, C] =
+    fa.asInstanceOf[F[W, EB, C]]
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpData.scala]:
```suggestion
```

Let's just typecast it for now.

File: zio-http/src/main/scala/zhttp/http/HttpData.scala [hunk 28]
```diff
@@ -1,10 +1,15 @@
...
 import io.netty.buffer.{ByteBuf, Unpooled}
+import zio.blocking.Blocking.Service.live.effectBlocking
 import zio.stream.ZStream
 import zio.{Chunk, NeedsEnv, UIO, ZIO}
 
+import java.io.RandomAccessFile
+import java.nio.ByteBuffer
+import java.nio.channels.FileChannel
 import java.nio.charset.Charset
+import scala.language.implicitConversions
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpData.scala]:
```suggestion
```

File: zio-http/src/main/scala/zhttp/http/HttpData.scala [hunk 29]
```diff
@@ -33,12 +38,31 @@ sealed trait HttpData[-R, +E] { self =>
...
+    fa.asInstanceOf[F[W, EB, C]]
+
+  def toByteBuf(implicit ev: E <:< Throwable): ZIO[R, Throwable, ByteBuf] = {
+    self match {
+      case HttpData.Text(text, charset)  => UIO(Unpooled.copiedBuffer(text, charset))
+      case HttpData.BinaryChunk(data)    => UIO(Unpooled.copiedBuffer(data.toArray))
+      case HttpData.BinaryByteBuf(data)  => UIO(data)
+      case HttpData.Empty                => UIO(Unpooled.EMPTY_BUFFER)
+      case HttpData.BinaryStream(stream) =>
+        stream.fold(Unpooled.compositeBuffer())((c, b) => c.addComponent(b))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpData.scala]:
```suggestion
        stream.asInstanceOf[ZStream[R, Throwable, ByteBuf]].fold(Unpooled.compositeBuffer())((c, b) => c.addComponent(b))
```

File: zio-http/src/main/scala/zhttp/http/Response.scala [hunk 30]
```diff
@@ -75,16 +76,27 @@ final case class Response[-R, +E] private (
       case HttpData.BinaryByteBuf(data) => data
       case HttpData.BinaryStream(_)     => null
       case HttpData.Empty               => Unpooled.EMPTY_BUFFER
+      case HttpData.File(_)             => null
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Response.scala]:
```suggestion
      case HttpData.File(_)             => 
        jHeaders.set(HttpHeaderNames.CONTENT_TYPE, Files.probeContentType(file.toPath))
        null
```

File: zio-http/src/main/scala/zhttp/http/Response.scala [hunk 31]
```diff
@@ -75,16 +76,27 @@ final case class Response[-R, +E] private (
...
+
+      // Set MIME type in the response headers. This is only relevant in case of File transfers as browsers use the MIME
+      // type, not the file extension, to determine how to process a URL.<a href="MSDN
+      // Doc">https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type</a>
+
+      self.data match {
+        case HttpData.File(file) =>
+          jHeaders.set(HttpHeaderNames.CONTENT_TYPE, Files.probeContentType(file.toPath))
+        case _                   => ()
+      }
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Response.scala]:
```suggestion
```

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 32]
```diff
@@ -0,0 +1,16 @@
+package zhttp.http
+
+import zio.test.Assertion.equalTo
+import zio.test.{DefaultRunnableSpec, ZSpec, assertM}
+
+import java.io.File
+
+object HttpDataSpec extends DefaultRunnableSpec {
+  // TODO : Add tests for othe HttpData types
+  override def spec: ZSpec[_root_.zio.test.environment.TestEnvironment, Any] =
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]:
```suggestion
  override def spec =
```

Unnecessary noise.

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 33]
```diff
@@ -0,0 +1,16 @@
...
+    suite("HttpDataSpec")(suite("Test toByteBuf")(testM("HttpData.File") {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]:
```suggestion
    suite("HttpDataSpec")(suite("Test toByteBuf")(testM("HttpData.fromFile") {
```

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 34]
```diff
@@ -129,6 +131,21 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
+        assertM(res)(equalTo("abc\nfoo"))
+      } +
+      testM("content-type header on file response") {
+        val file = new File(getClass.getResource("/TestFile.txt").getPath)
+        val res  =
+          Http
+            .fromFile(file)
+            .request()
+            .map(_.getHeaderValue(HttpHeaderNames.CONTENT_TYPE))
+            .map(_.getOrElse("Content type header not found."))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]:
```suggestion
            .requestHeaderValueByName(HttpHeaderNames.CONTENT_TYPE)            
```

**Comment by tusharmath** [comment]
@ashprakasan 
We don't need to test for large files specially for `toByteBuf` because that's only used in testing for now.

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 35]
```diff
@@ -129,6 +131,20 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
+        val res  = Http.fromFile(file).requestBodyAsString()
+        assertM(res)(equalTo("abc\nfoo"))
+      } +
+      testM("content-type header on file response") {
+        val file = new File(getClass.getResource("/TestFile.txt").getPath)
+        val res  =
+          Http
+            .fromFile(file)
+            .requestHeaderValueByName()(HttpHeaderNames.CONTENT_TYPE)
+            .map(_.getOrElse("Content type header not found."))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala:145]:
```suggestion
```

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 36]
```diff
@@ -129,6 +131,20 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
+        assertM(res)(equalTo("text/plain"))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala:146]:
```suggestion
        assertM(res)(isSome(equalTo("text/plain")))
```

**Review by tusharmath — APPROVED** [review]
